cilium/CFP-34841-bgp-route-learning.md

```diff
@@ -0,0 +1,118 @@
+
+Proposed here is a change to support route learning via BGP.
+
+## Motivation
+
+### User Stories
+
+![BGP Peering Diagram](./images/34841-bgp-route-learning.png)
+
+*As a cluster administrator, I wish to deploy Cilium on a node which connects to two or more routing domains.  The deployment is considered a success when Cilium is capable of learning non-overlapping routes using BGP from peers across each routing domain, and when Cilium is able to route outbound traffic to the correct routing domain.*
+
+
+https://github.com/cilium/cilium/pull/33035/
+
+
+## Goals
+
+* When Cilium is configured to use the Border Gateway Protocol (BGP) with one or more peers, routes advertised by those peers to Cilium are accepted, evaluated through BGP's Best Path Selection algorithm, and the resulting Best Path(s) installed into the node's dataplane.  The net result is that Cilium makes Internet Protocol (IP) routing decisions for Pod/Service sourced traffic based on routes learned via BGP.
+
+## Non-Goals
+
+* Making Cilium's BGP implementation Virtual Routing and Forwarding (VRF) instance aware.
+
+
+## Proposal
+
+### Overview
+
+### Route Programming
+
+Cilium supports two modes of routing -- [Encapsulation](https://docs.cilium.io/en/stable/network/concepts/routing/#encapsulation) and [Native](https://docs.cilium.io/en/stable/network/concepts/routing/#native-routing).
+
+#### Native Routing Mode
+
+When operating in Native Routing mode, the implementation appears straightforward.  A mechanism is required to synchronize routes learned from BGP, specifically Cilium's in-memory instance of GoBGP, to the Linux routing table.  For context, GoBGP may be deployed as a standalone daemon or instantiated directly through its Golang package.  Cilium instantiates directly.
+
+GoBGP on it's own is unable to install routes into the Linux routing table directly.  GoBGP's documentation [suggests](https://github.com/osrg/gobgp/blob/master/docs/sources/zebra.md) running an additional BGP daemon such as Quagga or FRR and establishing communication between the two.
+
+As the number of routes received, and rate of change may vary dramatically from one environment to another, it is the author's recommendation that one of the existing GoBGP supported-daemons be selected.  Under this model, an additional daemon will be deployed within Cilium's BGP-speaking pods.  For a Kubernetes-based deployment, these are the pods named `cilium-<hash>`.
+
+For Kubernetes deployments, where Cilium runs within a container, the mechanism to synchronize routes to the Linux routing table needs synchronize NOT to the container's routing table but rather the underlying node's routing table.  The author seeks guidance from the community on how to best approach this.
```

Comment:
We need to explore some areas 
  1. How should be graceful restart configured in such deployment.
  2. How do we protect the node and cluster from route hijacking.

Reply:
Re: route hijacking, do we have any protection today?

Reply:
Since we do not install BGP routes into kernel, this is not an issue for now.